### PR TITLE
Add generic D3D12 callback for pixel history

### PR DESCRIPTION
**Pull request by skarolewics**
This adds the next step toward D3D12 pixel history, which is mostly the generic action callback that is used for other replays later (similar to how Vulkan pixel history operates). There are some potential edge cases identified with TODOs but nothing that causes the functional test (to be checked in later) to fail.

### CODE REVIEW

File: renderdoc/driver/d3d12/d3d12_pixelhistory.cpp [hunk 1]
```diff
@@ -63,6 +63,31 @@ struct D3D12PixelHistoryResources
   D3D12Descriptor *dsDescriptor;
 };
 
+struct D3D12PixelHistoryCallbackInfo
+{
+  // Original image for which pixel history is requested.
+  WrappedID3D12Resource *targetImage;
+  D3D12_RESOURCE_DESC desc;
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_pixelhistory.cpp]: Minor nit - this would probably be better as `targetDesc` just to be clear what resource the description is of.

File: renderdoc/driver/d3d12/d3d12_pixelhistory.cpp [hunk 2]
```diff
@@ -246,6 +271,248 @@ void D3D12DebugManager::PixelHistoryCopyPixel(ID3D12GraphicsCommandListX *cmd,
...
+
+protected:
+  // Update the given scissor to just the pixel for which pixel history was requested.
+  void ScissorToPixel(const D3D12_VIEWPORT &view, D3D12_RECT &scissor)
+  {
+    float fx = (float)m_CallbackInfo.x;
+    float fy = (float)m_CallbackInfo.y;
+    float y_start = view.TopLeftY;
+    float y_end = view.TopLeftY + view.Height;
+    if(view.Height < 0)
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_pixelhistory.cpp:335]: I don't think negative viewport height is a thing on D3D12? This seems (harmlessly) left over from the vulkan implementation.

File: renderdoc/driver/d3d12/d3d12_pixelhistory.cpp [hunk 3]
```diff
@@ -246,6 +271,248 @@ void D3D12DebugManager::PixelHistoryCopyPixel(ID3D12GraphicsCommandListX *cmd,
...
+      pipeDesc.RasterizerState.CullMode = D3D12_CULL_MODE_NONE;
+      pipeDesc.RasterizerState.DepthClipEnable = FALSE;
+      pipeDesc.DepthStencilState.DepthBoundsTestEnable = FALSE;
+    }
+
+    // TODO: Get from callbackinfo/pixelhistoryresources?
+    pipeDesc.DSVFormat = DXGI_FORMAT_D32_FLOAT_S8X24_UINT;
+
+    // TODO: If the original depth buffer doesn't have stencil, this will not work as expected.
+    // Does this need to create a new depth/stencil to work correctly in that case?
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_pixelhistory.cpp]: Yes this will be one of the requirements of the pixel history, to patch in a stencil buffer as needed even to the extent of converting to a whole new depth buffer format.

**Review by baldurk — CHANGES_REQUESTED** [review]

File: renderdoc/driver/d3d12/d3d12_pixelhistory.cpp (hunk 2 above)
Comment (skarolewics) [reply renderdoc/driver/d3d12/d3d12_pixelhistory.cpp:335]: It was added in an Agility SDK release: https://devblogs.microsoft.com/directx/agility-sdk-1-602-0/

File: renderdoc/driver/d3d12/d3d12_pixelhistory.cpp (hunk 1 above)
Comment (skarolewics) [reply renderdoc/driver/d3d12/d3d12_pixelhistory.cpp]: Good point, renamed.

File: renderdoc/driver/d3d12/d3d12_pixelhistory.cpp (hunk 3 above)
Comment (skarolewics) [reply renderdoc/driver/d3d12/d3d12_pixelhistory.cpp]: Updated the TODO comment to indicate that it's definitely needed.
